### ISSUE TITLE
Assist: reorder fields

### ISSUE THREAD

**Issue by matklad** [rust-analyzer/rust-analyzer#3821]
This assist should be applicable to record literals and patterns, and should reorder fields in the same order as in the definition. Ie, it should be able to cleanup this:

![image](https://user-images.githubusercontent.com/1711539/78231726-87979080-74d3-11ea-9fa9-0f7a89be1298.png)

**Comment by Veetaha** [comment]
I wonder if there is some caveat in the order of evaluation of record literal fields initializers

**Comment by matklad** [comment]
Assists are not guaranteed to preserve semantics.

**Comment by JoshMcguigan** [comment]
I think this could get tricky:

```rust
struct Foo {
    a: String,
    b: String,
}

impl Foo {
    fn new() -> Self {
        let s = String::new();
        Self {
            b: s.clone(),
            a: s,
        }
    }
}
```
I'd think the output of an assist should always compile at least?

**Comment by matklad** [comment]
>I'd think the output of an assist should always compile at least?

Not necessary, assist is just a tool that applies some syntactic transformation which is helpful, it doesn't need to be complete or "correct". For example, in fill match arms, we deliberately use `{}` and not `todo!()` as a placeholder, because `{}` is easier to delete. 

The goal of assist is to be useful, and in your example I would be totally willing to apply the assist and move the clone call manually afterwards.

**Comment by geoffreycopin** [comment]
Hi,
First-time contributor here. I'd love to give this a try !

**Comment by eminence** [comment]
I believe this issue can be closed, as it was implemented in #3925
